src/odemis/driver/xt_client.py

```diff
@@ -2215,3 +2237,74 @@ def _setMultiBeamMode(self, multi_beam_mode):
         #     self.apertureIndex.value = current_aperture
         #     self.beamletIndex.value = current_beamlet
         return (self.parent.get_use_case() == 'MultiBeamTile')
+
+
+class MultiBeamDetector(Detector):
+    """
+    This is an extension of xt_client.Detector class. It overwrites the image acquisition
+    to work with image acquisition in XTToolkit.
+    """
+
+    def __init__(self, name, role, parent, channel_name, **kwargs):
+        """
+        channel_name (str): Name of one of the electron channels.
+        """
+        # The acquisition is based on a FSM that roughly looks like this:
+        # Event\State |    Stopped    |   Acquiring    | Receiving data |
+        #    START    | Ready for acq |        .       |       .        |
+        #    DATA     |       .       | Receiving data |       .        |
+        #    STOP     |       .       |     Stopped    |    Stopped     |
+        #    TERM     |     Final     |      Final     |     Final      |
+
+        Detector.__init__(self, name, role, parent=parent, channel_name=channel_name, **kwargs)
+
+    def stop_generate(self):
+        logging.debug("Stopping image acquisition")
+        # FIXME there has to be a better way to do this :(
+        # This doesn't work:
+        # self.parent.set_channel_state(self._channel_name, False)
+        server = Pyro5.api.Proxy("PYRO:Microscope@192.168.31.158:4242")
+        server.set_channel_state(self._channel_name, False)
+        if self.parent._scanner.blanker.value is None:
+            self.parent.blank_beam()
+        self._genmsg.put(GEN_STOP)
+
+    def _acquire(self):
+        """
+        Acquisition thread
+        Managed via the ._genmsg Queue
+        """
+        try:
+            while True:
+                # Wait until we have a start (or terminate) message
+                self._acq_wait_start()
+                logging.debug("Preparing acquisition")
+                while True:
+                    if self._acq_should_stop():
+                        break
+
+                    # Start a scan complete scan.
+                    self.parent.scan_image()
+                    # Check if the scan was stopped during acquiring.
+                    if self._acq_should_stop():
+                        break
+
+                    # Acquire the image
+                    # ======================================
+                    image = self.parent.get_latest_image(self._channel_name)
+
+                    md = self.parent._scanner._metadata.copy()
+                    md.update(self._metadata)
+                    md[model.MD_DWELL_TIME] = self.parent._scanner.dwellTime.value
+                    md[model.MD_ROTATION] = self.parent._scanner.rotation.value
+
+                    da = DataArray(image, md)
+                    logging.debug("Notify dataflow with new image.")
+                    self.data.notify(da)
```

Comment:
Then this can also go into one if statement like:

> if self._acq_should_stop():
>                     # Acquire the image
>                     # ======================================
>                     image = self.parent.get_latest_image(self._channel_name)

I don't particularly like the break statements in this context.